### ISSUE TITLE
[java] use Pattern variable: Variable 'collection' can be replaced with pattern variable

### ISSUE THREAD

**Issue by Pankraz76** [pmd/pmd#5761]
**UsePatternVariable**

**Proposed Category:** One of [Best Practices | Code Style | Design ]

**Description:**

Variable 'n' can be replaced with pattern variable 

<img width="763" alt="Image" src="https://github.com/user-attachments/assets/8f30f97a-7b09-4944-9b39-da30e9930ecb" />

**Code Sample:** This should include code, that should be flagged by the rule. If possible, the "correct" code
according to this new rule should also be demonstrated.

before:
```java
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }

        if (!(obj instanceof FileModelSource)) {
            return false;
        }
        FileModelSource other = (FileModelSource) obj;
        return getFile().equals(other.getFile());
    }
```
after:
```java
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }

        if (!(obj instanceof FileModelSource other)) {
            return false;
        }
        return getFile().equals(other.getFile());
    }
```

before:
```java
    public static void serialize(Properties props, String key, Object value) {
        if (value != null && value.getClass().isArray()) {
            props.setProperty(key, Integer.toString(Array.getLength(value)));
            for (int i = Array.getLength(value) - 1; i >= 0; i--) {
                serialize(props, key + "." + i, Array.get(value, i));
            }
        } else if (value instanceof Collection) {
            Collection collection = (Collection) value;
            props.setProperty(key, Integer.toString(collection.size()));
            int i = 0;
            for (Iterator it = collection.iterator(); it.hasNext(); i++) {
                serialize(props, key + "." + i, it.next());
            }
        } else if (value instanceof Map) {
            Map map = (Map) value;
            props.setProperty(key, Integer.toString(map.size()));
            @SuppressWarnings("checkstyle:UnusedLocalVariable")
            int i = 0;
            for (Iterator it = map.keySet().iterator(); it.hasNext(); i++) {
                Object k = it.next();
                Object v = map.get(k);
                serialize(props, key + "." + k, v);
            }
        } else if (value instanceof PlexusConfiguration) {
            PlexusConfiguration config = (PlexusConfiguration) value;

            String val = config.getValue(null);
            if (val != null) {
                props.setProperty(key + ".value", val);
            }

            String[] attributes = config.getAttributeNames();
            props.setProperty(key + ".attributes", Integer.toString(attributes.length));
            for (int i = attributes.length - 1; i >= 0; i--) {
                props.setProperty(key + ".attributes." + attributes[i], config.getAttribute(attributes[i], ""));
            }

            PlexusConfiguration children[] = config.getChildren();
            props.setProperty(key + ".children", Integer.toString(children.length));
            Map<String, Integer> indices = new HashMap<>();
            for (PlexusConfiguration child : children) {
                String name = child.getName();
                Integer index = indices.get(name);
                if (index == null) {
                    index = 0;
                }
                serialize(props, key + ".children." + name + "." + index, child);
                indices.put(name, index + 1);
            }
        } else if (value instanceof Date) {
            props.setProperty(key, new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format((Date) value));
        } else if (value != null) {
            props.setProperty(key, value.toString());
        }
    }
```

after:
```java
    public static void serialize(Properties props, String key, Object value) {
        if (value != null && value.getClass().isArray()) {
            props.setProperty(key, Integer.toString(Array.getLength(value)));
            for (int i = Array.getLength(value) - 1; i >= 0; i--) {
                serialize(props, key + "." + i, Array.get(value, i));
            }
        } else if (value instanceof Collection collection) {
            props.setProperty(key, Integer.toString(collection.size()));
            int i = 0;
            for (Iterator it = collection.iterator(); it.hasNext(); i++) {
                serialize(props, key + "." + i, it.next());
            }
        } else if (value instanceof Map map) {
            props.setProperty(key, Integer.toString(map.size()));
            @SuppressWarnings("checkstyle:UnusedLocalVariable")
            int i = 0;
            for (Iterator it = map.keySet().iterator(); it.hasNext(); i++) {
                Object k = it.next();
                Object v = map.get(k);
                serialize(props, key + "." + k, v);
            }
        } else if (value instanceof PlexusConfiguration config) {

            String val = config.getValue(null);
            if (val != null) {
                props.setProperty(key + ".value", val);
            }

            String[] attributes = config.getAttributeNames();
            props.setProperty(key + ".attributes", Integer.toString(attributes.length));
            for (int i = attributes.length - 1; i >= 0; i--) {
                props.setProperty(key + ".attributes." + attributes[i], config.getAttribute(attributes[i], ""));
            }

            PlexusConfiguration children[] = config.getChildren();
            props.setProperty(key + ".children", Integer.toString(children.length));
            Map<String, Integer> indices = new HashMap<>();
            for (PlexusConfiguration child : children) {
                String name = child.getName();
                Integer index = indices.get(name);
                if (index == null) {
                    index = 0;
                }
                serialize(props, key + ".children." + name + "." + index, child);
                indices.put(name, index + 1);
            }
        } else if (value instanceof Date) {
            props.setProperty(key, new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format((Date) value));
        } else if (value != null) {
            props.setProperty(key, value.toString());
        }
    }

```

```console
+        } else if (value instanceof Collection collection) {
-        } else if (value instanceof Collection) {
-            Collection collection = (Collection) value;
```

**Possible Properties:**

*   Should this rule be customizable via properties?

maybe for exclusions, for happy path not.

- [use Pattern variable: Variable 'collection' can be replaced with pattern variable](https://github.com/apache/maven/pull/2385)

**Comment by Pankraz76** [comment]
I actually did not assume this code to work, as the scope of the pattern variable was expected to be limited to the block `{}`, but that's not the case.

**Comment by adangel** [comment]
Thanks for the rule suggestion. This is a rule that helps to update an existing codebase to use newer java feature. Be aware, that this rule can only be applied, if the project uses Java 16 or later. Rules in PMD allow to declare a minimum language version, which should be used in that case.

The feature of Java you are referring to is called [JEP 394: Pattern Matching for instanceof](https://openjdk.org/jeps/394). The scope of the variables is a new thing called "flow scoping".

We have some [Rule Guidelines](https://docs.pmd-code.org/latest/pmd_devdocs_major_rule_guidelines.html) which we try to apply. New rules are good candidates to create better rule names and better descriptions (changing the name of a rule later is possible but not as easy as getting it right in the first place).
Maybe you can check this page and reevaluate whether there is a better name. Maybe "Use Pattern Matching"?

The examples are good, but IMHO too detailed. To explain, what the rule does, two simple cases are enough:

```java
// instead of
void example1_before(Object o) {
    if (o instanceof String) {
        String s = (String) o;
        System.out.println("Length: " + s.length());
    }
// with pattern matching
void example1_after(Object o) {

    if (o instanceof String s) {
        System.out.println("Length: " + s.length());
    }
}

// instead of
void example2_before(Object o) {
    if (!o instanceof String)) {
        return;
    }
    String s = (String) o;
    System.out.println("Length: " + s.length());
}

// with pattern matching
void example2_after(Object o) {
    if (!o instanceof String s)) {
        return;
    }
    System.out.println("Length: " + s.length());
}
```

Your more detailed examples might be useful for the tests.

I'd say, we don't need properties for the rule. At least, not at the beginning. We should start simple.

Do you maybe want to give it a try and implement this rule?

**Comment by Pankraz76** [comment]
Appreciate the feedback, indeed its to complex as c&p from real world example.
Yes will give it a try.
